src/averageAges.js

```diff
@@ -1,61 +1,57 @@
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
- * @param {object[]} people
- * @param {number} century - optional
- *
- * @return {number}
- */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  let mens = [];
+
+  century ? (
+    mens = people.filter(person => person.sex === 'm'
+    && Math.ceil(person.died / 100) === century)
+  ) : mens = people.filter(person => person.sex === 'm');
+
+  const mensAvgAge = mens
+    .map(men => men.died - men.born)
+    .reduce((totalYears, menAge) => totalYears + menAge, 0);
+
+  return Math.round(mensAvgAge / mens.length * 100) / 100 || 0;
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  let womans = [];
+
+  withChildren ? (
+    womans = people
+      .filter(person => person.sex === 'f'
+    && people.some(woman => person.name === woman.mother))
+  ) : (
+    womans = people.filter(person => person.sex === 'f')
+  );
+
+  const womansAvgAge = womans
+    .map(woman => woman.died - woman.born)
+    .reduce((totalYears, womanAge) => totalYears + womanAge, 0);
+
+  return Math.round(womansAvgAge / womans.length * 100) / 100 || 0;
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  let childs = [];
+
+  onlyWithSon ? (
+    childs = people
+      .filter(child => child.sex === 'm'
+        && people.find(woman => woman.name === child.mother))
+  ) : (
+    childs = people
+      .filter(child => people.find(woman =>
+        woman.name === child.mother))
+  );
+
+  const childsAvgAge = childs
+    .reduce((total, child) =>
+      total + (child.born - people.find(woman =>
+        woman.name === child.mother).born), 0);
+
+  return Math.round(childsAvgAge / childs.length * 100) / 100 || 0;
```

Comment:
maybe '/childs.length' better to do after reduce? but it's my point of view)

Reply:
Yeah, it one of the ways but it'll be harder to read this code by other in my mind